miasm2/ir/ssa.py

```diff
@@ -0,0 +1,279 @@
+
+        :param assignblk: assignblock
+        :return: sorted list of expressions
+        """
+        instructions = []
+        for dst in assignblk:
+            # dst = src
+            aff = m2_expr.ExprAff(dst, assignblk[dst])
+            # insert memory expression into start of list
+            if isinstance(dst, m2_expr.ExprMem):
+                instructions.insert(0, aff)
+            else:
+                instructions.append(aff)
+
+        return instructions
+
+    @staticmethod
+    def _convert_block(ib, l):
+        """
+        Transforms an IRA block inplace into SSA
+        :param ib: IRA block to be transformed
+        :param l: list of SSA expressions
+        """
+        # iterator over SSA expressions
+        ssa_iter = iter(l)
+        # walk over IR blocks' assignblocks
+        for index, assignblk in enumerate(ib.irs):
+            # list of instructions
+            instructions = []
+            # insert SSA instructions
+            for dst in assignblk:
+                instructions.append(ssa_iter.next())
+            # replace instructions of assignblock in IRA block
+            ib.irs[index] = AssignBlock(instructions)
+
+    def _copy_block(self, block):
+        """
+        Returns a copy on an IRA block
+        :param block: asm_label
+        :return: IRA block
+        """
+        # retrieve IRA block
+        ib = self.ira.get_bloc(block)
+
+        # copy IRA block
+        irs = [assignblk.copy() for assignblk in ib.irs]
+        ib_ssa = irbloc(ib.label, irs)
+
+        # set next block
+        ib_ssa.dst = ib.dst.copy()
+
+        # add to SSA blocks dict
+        self.blocks.update({ib_ssa.label: ib_ssa})
+
+        return ib_ssa
+
+    def _rename_expressions(self, block):
+        """
+        Transforms variables and expressions
+        of an IRA block into SSA.
+
+        IR representations of an assembly instruction are evaluated
+        in parallel. Thus, RHS and LHS instructions will be performed
+        separately.
+        :param block: IRA block label
+        """
+        # list of IRA block's SSA expressions
+        ssa_expressions_block = []
+
+        # retrieve IRA block
+        ib = self.get_block(block)
+
+        # iterate block's IR expressions
+        for assignblk in ib.irs:
+            # list of parallel instructions
+            instructions = self._parallel_instructions(assignblk)
+            # list for transformed RHS expressions
+            rhs = []
+
+            # transform RHS
+            for e in instructions:
+                src = e.src
+                src_ssa = self._transform_expression_rhs(src)
+                # save transformed RHS
+                rhs.append(src_ssa)
+
+            # transform LHS
+            for e in instructions:
+                dst = e.dst
+                dst_ssa = self._transform_expression_lhs(dst)
+
+                # retrieve corresponding RHS expression
+                src_ssa = rhs.pop(0)
```

Comment:
List are not very effective when [used as queue](https://docs.python.org/2/tutorial/datastructures.html#using-lists-as-queues). You can use `deque` instead, which is more effective